### PR TITLE
[UserStory] QT4BJ extension has its own panel in BlueJ Preferences>Extensions #112

### PR DESCRIPTION
**Pull request by holsappled**
Created Preferences class and associated acceptance test as an AcceptanceTest.md.

### CODE REVIEW

File: AcceptanceTest.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# AcceptanceTest for QualityToolsExtension.jar
```
Comment (jody) [inline AcceptanceTest.md]: Is there **exactly one and only one** acceptance test?  Is a new file needed for an additional test?

Is the intent that this only apply to a file named _QualityToolsExtension.jar_?  Does it need to be rewritten or replaced if the name of the jar file evolves during the lifetime of the product?

File: AcceptanceTest.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+Due to the difficulty of JUnit tests developers created the following acceptance test to verify the start up functionality of the BlueJ application.
```
Comment (jody) [inline AcceptanceTest.md]: What is meant by "Due to the difficulty of JUnit tests"?  What is the intent of the opening phrase of this document?

File: AcceptanceTest.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# AcceptanceTest for QualityToolsExtension.jar
+
+Due to the difficulty of JUnit tests developers created the following acceptance test to verify the start up functionality of the BlueJ application. 
+
+
+# Steps 
+
+1) Gradle build the project with JDK version 11
```
Comment (jody) [inline AcceptanceTest.md]: As I progress through this document, I begin to understand that this is intended more as a "one off" acceptance criterion than something useful for longer duration of product development and maintenance.

If this is the case, then it should appear as "Acceptance Criteria" in the associated issue, rather than as a document in the repository.

Consider copying the verification steps to the appropriate issue(s) and deleting this document.

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java [hunk 4]
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
...
 }
+
+/**
+ * Create a Preference Panel located in the BlueJ application
+ * BlueJ -> Preferences -> Extensions Tab
+ */
+class Preferences implements PreferenceGenerator
+{
+    private JPanel myPanel;
+    private JTextField color;
```
Comment (jody) [inline src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: What is the intent of this field?  It seems irrelevant.

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java [hunk 5]
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
...
     }
 }
+
+/**
+ * Create a Preference Panel located in the BlueJ application
+ * BlueJ -> Preferences -> Extensions Tab
+ */
+class Preferences implements PreferenceGenerator
+{
+    private JPanel myPanel;
```
Comment (jody) [inline src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: All fields and methods should be commented, using Javadoc-style comments.

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java [hunk 6]
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
         }
     }
 }
+
+/**
+ * Create a Preference Panel located in the BlueJ application
```
Comment (jody) [inline src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: This comment seems to indicate that this class determines the location of the panel.  That doesn't seem to be the actual way this is working.

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java [hunk 7]
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
...
+    public static final String PROFILE_LABEL="Quality-Analysis-Tools";
+
+    // Construct the panel, and initialise it from any stored values
+    public Preferences(BlueJ bluej)
+    {
+        this.bluej = bluej;
+        myPanel = new JPanel();
+        myPanel.add (new JLabel ("Quality Analysis Tools"));
+        color = new JTextField (40);
+        myPanel.add (color);
```
Comment (jody) [inline src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: What is "color"?

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java [hunk 8]
```diff
@@ -91,3 +99,42 @@ public URL getURL() {
...
+        // Load the default value
```
Comment (jody) [inline src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: What is meant by "default value"?

**Review by jody — CHANGES_REQUESTED** [review]
Please have multiple members of your team perform a code review and attest to the quality and conformance to conventions.

File: AcceptanceTest.md (hunk 1 above)
Comment (shadycloud) [reply AcceptanceTest.md]: 1) Within the ```AcceptanceTest.md``` file.  There are two acceptance tests.  First test is to verify ```QualityTools``` Extension displays.  The second acceptance test being Quality Tools Extension displays from Preferences “**Extensions Tab**”.

2) It seems our organization has decided to implement that name for our production jar files as in User story #149 and #138 which could be confusing. We are proposing that we name it ```ProductAcceptanceTest.md```

Comment (shadycloud) [reply AcceptanceTest.md]: It seems the MetroCS organization plans on the future implementation of semantic versioning in our production jar files as detailed in User story #149 and #138. However, slight modifications will need to be updated ``` "rewritten or replaced" ``` in the file name and body content until a more permanent strategy can be developed to automagically handle semantic versioning of file names. https://semver.org/

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java (hunk 4 above)
Comment (holsappled) [reply src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: Field is needed to create JTextField from the Swing API.  

After discussing with the group, having a JTextField displayed in the quality tools extension tab serves no purpose.  The intent was to match the #112 user story "picture display"

new commit will remove field

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java (hunk 7 above)
Comment (holsappled) [reply src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: We will remove "color" object as it was used to store the JTextField.  There is no user input required for this user story.  Will remove lines of code associated with "color".  

For reference and to answer your question see the information below.  

`color = new JTextField (40);`
API recommends Note: 
"We encourage you to specify the number of columns for each text field. If you do not specify the number of columns or a preferred size, then the field's preferred size changes whenever the text changes, which can result in unwanted layout updates."

https://docs.oracle.com/javase/tutorial/uiswing/components/textfield.html

File: AcceptanceTest.md (hunk 2 above)
Comment (holsappled) [reply AcceptanceTest.md]: Invoking the BlueJ proxy object to create a JUnit test.

File: AcceptanceTest.md (hunk 3 above)
Comment (holsappled) [reply AcceptanceTest.md]: After discussing with the group, having a JTextField displayed in the quality tools extension tab serves no purpose. The intent was to match the #112 user story "picture display"

**Comment by holsappled** [comment]
New commits will reflect BlueJ application under BlueJ->Preferences->Extensions Tab

Screen Shot 2019-12-04 at 6 38 37 AM

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java (hunk 4 above)
Comment (jody) [reply src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: > The intent was to match the #112 user story "picture display"

Matching that with something relevant to the product, such as, "Quality Analysis Tools", would have been fine.  "color" just seems irrelevant.

File: src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java (hunk 7 above)
Comment (jody) [reply src/main/java/edu/msudenver/bluejext/qualitytools/QualityToolsExtension.java]: Just a note that it is less about "user input" and more about the choice of "color" in particular as having no obvious relevance and likely being an unintended artifact of a copy-paste of work done by others.

**Review by jody — CHANGES_REQUESTED** [review]
This code doesn't appear to follow the coding convention guidelines for the project.
See: [Coding Convention Guide](https://github.com/MetroCS/QualityToolsForBlueJ/blob/master/DevelopmentConventions.md#Coding-Convention-Guide)

**Review by jody — APPROVED** [review]
Will approve to close this out. Need to open bug report to fix formatting (misplaced '{', etc.) to adhere to coding convention.
